Mathlib/CategoryTheory/Localization/StructuredArrow.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.HomEquiv
+import Mathlib.CategoryTheory.Localization.Opposite
+import Mathlib.CategoryTheory.Comma.StructuredArrow
+
+/-!
+# Induction principles for structured and costructured arrows
+
+Assume that `L : C ⥤ D` is a localization functor for `W : MorphismProperty C`.
+Given `X : C` and a predicate `P` on `StructuredArrow (L.obj X) L`, we obtain
+the lemma `Localization.induction_structuredArrow` which shows that `P` holds for
+all structured arrows if `P` holds for the identity map `𝟙 (L.obj X)`,
+if `P` is stable by post-composition with `L.map f` for any `f`
+and if `P` is stable by post-composition with the inverse of `L.map w` when `W w`.
+
+We obtain a similar lemma `Localization.induction_costructuredArrow` for
+costructured arrows.
+
+-/
+
+namespace CategoryTheory
+
+open Opposite
+
+variable {C D D' : Type*} [Category C] [Category D] [Category D']
+
+namespace Localization
+
+section
+
+variable (W : MorphismProperty C) (L : C ⥤ D) (L' : C ⥤ D')
+  [L.IsLocalization W] [L'.IsLocalization W] {X : C}
+
+/-- The bijection `StructuredArrow (L.obj X) L ≃ StructuredArrow (L'.obj X) L'`
+when `L` and `L'` are two localization functors for the same class of morphisms. -/
+@[simps]
+noncomputable def structuredArrowEquiv :
+    StructuredArrow (L.obj X) L ≃ StructuredArrow (L'.obj X) L' where
+  toFun f := StructuredArrow.mk (homEquiv W L L' f.hom)
+  invFun f := StructuredArrow.mk (homEquiv W L' L f.hom)
+  left_inv f := by
+    obtain ⟨Y, f, rfl⟩ := f.mk_surjective
+    dsimp
+    rw [← homEquiv_symm_apply, Equiv.symm_apply_apply]
+  right_inv f := by
+    obtain ⟨Y, f, rfl⟩ := f.mk_surjective
+    dsimp
+    rw [← homEquiv_symm_apply, Equiv.symm_apply_apply]
+
+end
+
+section
+
+variable (W : MorphismProperty C) {X : C}
+  (P : StructuredArrow (W.Q.obj X) W.Q → Prop)
+  (hP₀ : P (StructuredArrow.mk (𝟙 (W.Q.obj X))))
+  (hP₁ : ∀ ⦃Y₁ Y₂ : C⦄ (f : Y₁ ⟶ Y₂) (φ : W.Q.obj X ⟶ W.Q.obj Y₁),
+    P (StructuredArrow.mk φ) → P (StructuredArrow.mk (φ ≫ W.Q.map f)))
+  (hP₂ : ∀ ⦃Y₁ Y₂ : C⦄ (w : Y₁ ⟶ Y₂) (hw : W w) (φ : W.Q.obj X ⟶ W.Q.obj Y₂),
+    P (StructuredArrow.mk φ) → P (StructuredArrow.mk (φ ≫ (isoOfHom W.Q W w hw).inv)))
+
+open Construction in
+lemma induction_structuredArrow' (g : StructuredArrow (W.Q.obj X) W.Q) : P g := by
```

Comment:
Am I correct that `induction_structuredArrow` is strictly more general? If so, maybe `private` this or indicate that this is an auxiliary lemma in the docstring.

Reply:
Yes, I have made it private.